src/Microsoft.AspNetCore.Sockets/Internal/AuthorizeHelper.cs

```diff
@@ -12,53 +13,62 @@ namespace Microsoft.AspNetCore.Sockets.Internal
-        public static async Task<bool> AuthorizeAsync(HttpContext context, AuthorizationPolicy policy)
+        public static async Task<bool> AuthorizeAsync(HttpContext context, IList<string> policies)
         {
-            if (policy != null)
+            if (policies.Count == 0)
             {
-                if (policy.AuthenticationSchemes != null && policy.AuthenticationSchemes.Count > 0)
-                {
-                    ClaimsPrincipal newPrincipal = null;
-                    foreach (var scheme in policy.AuthenticationSchemes)
-                    {
-                        var result = await context.Authentication.AuthenticateAsync(scheme);
-                        if (result != null)
-                        {
-                            newPrincipal = SecurityHelper.MergeUserPrincipal(newPrincipal, result);
-                        }
-                    }
-
-                    if (newPrincipal == null)
-                    {
-                        newPrincipal = new ClaimsPrincipal(new ClaimsIdentity());
-                    }
+                return true;
+            }
 
-                    context.User = newPrincipal;
-                }
+            var policyProvider = context.RequestServices.GetRequiredService<IAuthorizationPolicyProvider>();
+            var authorizeData = new List<IAuthorizeData>();
 
-                var authService = context.RequestServices.GetRequiredService<IAuthorizationService>();
-                if (await authService.AuthorizeAsync(context.User, context, policy))
-                {
-                    return true;
-                }
+            foreach (var policy in policies)
+            {
+                authorizeData.Add(new AuthorizeAttribute(policy));
+            }
 
-                // Challenge
-                if (policy.AuthenticationSchemes != null && policy.AuthenticationSchemes.Count > 0)
+            var authorizePolicy = await AuthorizationPolicy.CombineAsync(policyProvider, authorizeData);
+            if (authorizePolicy.AuthenticationSchemes != null && authorizePolicy.AuthenticationSchemes.Count > 0)
+            {
+                ClaimsPrincipal newPrincipal = null;
+                foreach (var scheme in authorizePolicy.AuthenticationSchemes)
                 {
-                    foreach (var scheme in policy.AuthenticationSchemes)
+                    var result = await context.Authentication.AuthenticateAsync(scheme);
+                    if (result != null)
                     {
-                        await context.Authentication.ChallengeAsync(scheme, properties: null);
+                        newPrincipal = SecurityHelper.MergeUserPrincipal(newPrincipal, result);
                     }
                 }
-                else
+
+                if (newPrincipal == null)
                 {
-                    await context.Authentication.ChallengeAsync(properties: null);
+                    newPrincipal = new ClaimsPrincipal(new ClaimsIdentity());
```

Comment:
Its just a minor optimization, we don't want to allocate it unless none of the `AuthenticationSchemes` could be authenticated.